design/hardware-classification-controller/support-for-new-parameters-hwcc.md

```diff
@@ -0,0 +1,192 @@
+We would like to add more parameters which helps user to
+identify the ideal host. This feature helps user to avoid runtime failures
+and also increases the performance for workload deployments.
+
+### Goals
+
+To make classification more useful by incorporating additional parameters
+to get the matching host. It will help user to specify more rules by
+introducing new parameters.
+
+We will carry out testing of HWCC on Dell hardware.
+
+## Proposal
+
+We are submitting proposal based on the issue raised in HWCC:
+[Issue 56](https://github.com/metal3-io/hardware-classification-controller/issues/56)
+
+* CPU architecture will be useful if the user wants to have 32-bit
+  or 64-bit processor as per his requirements.
+
+* New parameters such as Firmware and SystemVendor will add
+  advantage to identify matched hosts.
+  System Vendor will be useful to add as a new parameter if the
+  user is looking for particular vendor and their hardware for his
+  application to support or deploy on.
+  Firmware will be useful to add as a new parameter if user is looking
+  for specific features in particular releases or versions.
+
+### Investigation Details
+
+Add support for new parameters:
+
+1. `hardwareCharacteristics.cpu.arch`
+
+   The new `arch` field will be compared against the host
+   `status.hardwareDetails.cpu.arch` field for an exact string match.
+
+   User can provide `arch` type based on requirement.
+
+   Investigation Details: Below is sample introspection data for CPU.
+   Types of architecture are 32-bit (x86) and 64-bit
+   (x86-64, IA64, and AMD64).
+
+   ```yaml
+   hardware:
+       cpu:
+          arch: "x86_64"
+   ```
+
+1. `hardwareCharacteristics.firmware.bios`
+
+   Firmware is new parameter to be introduced in HardwareClassification
+   which contains BIOS information. User can specify BIOS requirements
+   by providing `hardwareCharacteristics.firmware.bios.vendor` and
+   `hardwareCharacteristics.firmware.bios.version`.
+
+   The classification of `vendor` will be done using exact string
+   match of the value provided by user in profile.
+
+   The classification of `version` will be done by checking if host
+   `version` is greater than equal to `minorVersion` and less than
+   equal to `majorVersion`. `minorVersion` and `majorVersion` are
+   provided by user in profile.
+
+   Investigation details: Below is sample introspection data for firmware.
+   `BIOS` field under firmware have two fields, `vendor` and `version`.
+
+    ```yaml
+    hardware:
+        firmware:
+             bios:
+               date: "10/17/2018"
+               vendor: "Dell Inc."
+               version: "1.5.6"
+    ```
+
+1. `hardwareCharacteristics.systemVendor`
+
+   SystemVendor is new parameter to be introduced which helps user to
+   classify hosts of certain manufacturer and product. User needs to
+   provide `hardwareCharacteristics.systemVendor.manufacturer` and
+   `hardwareCharacteristics.systemVendor.productName` in yaml.
```

Comment:
Is it really useful to require `productName`? Also this duplicates `vendor` in `firmware`, why do we need both?

Reply:
All the fields are optional.
For example :  if user decide to fetch host only on SystemVendor information. Then there is no need to provide vendor in Firmware..

There is a grammatical mistake. Updating it as "User **can** provide `hardwareCharacteristics.systemVendor.manufacturer` and
   `hardwareCharacteristics.systemVendor.productName` in yaml".

Reply:
The host API has both of these fields as separate, so it makes sense to do the same here. I thought that the host schema was designed to match the inspection data. Does the inspection data not have those as separate fields?

Is the firmware vendor always the same as the system vendor?

Reply:
Removing Vendor from Bios because value under `systemVendor.Manufacturer` and `Bios.Vendor` are same `Dell Inc.` for dell hardware.

Reply:
> Removing Vendor from Bios because value under `systemVendor.Manufacturer` and `Bios.Vendor` are same `Dell Inc.` for dell hardware.

We support more than Dell hardware, though, right? Are those fields always the same for all vendors?

We originally used separate fields in the host API because Ironic reports the values separately. They are semantically different, so even if the values are the same I think it is useful to keep both fields.

Reply:
Yes, it is possible those fields may differ for other vendors.
Reverting the change.